RFS/src/main/java/com/rfs/common/LuceneDocumentsReader.java

```diff
@@ -17,58 +21,116 @@
 @RequiredArgsConstructor
 @Slf4j
 public class LuceneDocumentsReader {
+    public static Function<Path, LuceneDocumentsReader> getFactory(boolean softDeletesPossible, String softDeletesField) {
+        return path -> new LuceneDocumentsReader(path, softDeletesPossible, softDeletesField);
+    }
     public static final int NUM_DOCUMENTS_BUFFERED = 1024;
+
     protected final Path indexDirectoryPath;
+    protected final boolean softDeletesPossible;
+    protected final String softDeletesField;
+
+    /**
+     * There are a variety of states the documents in our Lucene Index can be in; this method extracts those documents
+     * that would be considered "live" from the ElasticSearch/OpenSearch perspective.
+     *
+     * For context, when ElasticSearch/OpenSearch deletes a document, it doesn't actually remove it from the Lucene Index.
+     * Instead, what happens is that the document is marked as "deleted" in the Lucene Index, but it is still present in the
+     * Lucene segment on disk.  The next time that segment is merged, the deleted documents are removed from the Lucene Index.
+     * A similar thing happens when a document is updated; the old document is marked as "deleted" and a new document is
+     * added in a new Lucene segment.  This means that from an ES/OS perspective, you could have a single document that has
+     * been created, deleted, recreated, updated, etc. multiple times and only a single version of the doc would exist when
+     * you queried the ES/OS Index - but every single iteration of that doc might still exist in the Lucene Segments on disk,
+     * all of which have the same _id (from the ES/OS perspective).
+     *
+     * Additionally, Elasticsearch 7 introduced a feature called "soft deletes" which allows you to mark a document as
+     * "deleted" in the Lucene Index without actually removing it from the Lucene Index.  This works by having the
+     * application writing the Lucene Index define a field that is used to mark a document as "soft deleted" or not.  When
+     * a document is marked as "soft deleted", it is not returned in search results, but it is still present in the Lucene
+     * Index.  The status of whether any given document is "soft deleted" or not is stored in the Lucene Index itself.  By
```

Comment:
The documentation here not very helpful [1], but I think the idea is that the soft deletes are only "resolved"/"closed" once the Elasticsearch cluster is confident that all replicas have successfully resolved them.  This provides an additional layer of safety beyond the translog on the primary so that if a replica node loses connectivity to the cluster, it can successfully resolve the deletes that have been performed while it was diconnected.  In that scenario, the primary may have already performed a Lucene Merge and truncated its translog, forcing the replica to pull a full new copy of the shard when it comes back into the cluster instead of being able to just replay the operations it missed.  Soft-deletes appears like an optimization to make that expensive operation less likely to be necessary.

[1] https://www.elastic.co/guide/en/elasticsearch/reference/7.10/index-modules-history-retention.html